sympy/functions/special/beta_functions.py

```diff
@@ -118,3 +132,126 @@ def _eval_conjugate(self):
+
+###############################################################################
+########################## INCOMPLETE BETA FUNCTION ###########################
+###############################################################################
+
+class betainc(Function):
+    r"""
+    The Generalized Incomplete Beta function is defined as
+
+    .. math::
+        \mathrm{B}_{(x_1, x_2)}(a, b) = \int_{x_1}^{x_2} t^{a - 1} (1 - t)^{b - 1} dt
+
+    The Beta function is a special case of the
+    Generalized Incomplete Beta function :
+
+    .. math:: \mathrm{B}(a, b) = \mathrm{B}_{(0, 1)}(a, b)
+
+    The Generalized Regularized Incomplete Beta function is given by
+
+    .. math::
+        \mathrm{I}_{(x_1, x_2)}(a, b) = \frac{\mathrm{B}_{(x_1, x_2)}(a, b)}{\mathrm{B}(a, b)}
+
+    The Regularized Incomplete Beta function is the cumulative distribution
+    function of the beta distribution.
+
+    Examples
+    ========
+
+    >>> from sympy import I, pi, E, symbols
+    >>> from sympy import betainc, conjugate
+    >>> a, b, x, x1, x2 = symbols('a b x x1 x2')
+
+    The Generalized Incomplete Beta function is
+    given by
+
+    >>> betainc(a, b, x1, x2)
+    betainc(a, b, x1, x2, 0)
+
+    The Generalized Regularized Incomplete Beta function
+    is given by
+
+    >>> betainc(a, b, x1, x2, regularized=True)
+    betainc(a, b, x1, x2, 1)
+
+    The Incomplete Beta function is a special case
+    of the Generalized Incomplete Beta Function.
+    It can be obtained as follows:
+
+    >>> betainc(a, b, 0, x)
+    betainc(a, b, 0, x, 0)
+
+    The Regularized Incomplete Beta function is a special case
+    of the Generalized Regularized Incomplete Beta Function.
+    It can be obtained as follows:
+
+    >>> betainc(a, b, 0, x, regularized=True)
+    betainc(a, b, 0, x, 1)
+
+    The Beta function obeys the mirror symmetry:
+
+    >>> conjugate(betainc(a, b, x1, x2))
+    betainc(conjugate(a), conjugate(b), conjugate(x1), conjugate(x2), 0)
+
+    We can numerically evaluate the Incomplete Beta function
+    to arbitrary precision on the whole complex plane:
+
+    >>> from sympy import betainc
+    >>> betainc(2, 3, 4, 5).evalf(10)
+    56.08333333
+    >>> betainc(0.75, 1 - 4*I, 0, 2 + 3*I).evalf(25)
+    0.2241657956955709603655887 + 0.3619619242700451992411724*I
+    >>> betainc(pi, E, 0, 1, regularized=True).evalf(5)
+    1.00000
+
+    See Also
+    ========
+
+    beta: Beta function
+    hyper: Generalized Hypergeometric function
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Beta_function#Incomplete_beta_function
+    .. [2] https://dlmf.nist.gov/8.17
+    .. [3] https://functions.wolfram.com/GammaBetaErf/Beta4/
+    .. [4] https://functions.wolfram.com/GammaBetaErf/BetaRegularized4/02/
+
+    """
+    nargs = 5
+    unbranched = True
+
+    def __new__(cls, a, b, x1, x2, regularized=False):
+        return Function.__new__(cls, a, b, x1, x2, int(regularized == True))
```

Comment:
Storing a boolean flag like this is a bad idea. This should just be two separate classes.

Reply:
Yes, I felt that too, but I would have to pass in a boolean flag anyway to the mpmath function. Even if I had another class, say `regularized_betainc`, I would have to pass in the boolean flag 1. Is there a better way to do this?

Reply:
Passing a boolean flag in just one method is fine. At the moment most methods of the class look like:
```
if self.flag:
    # one thing
else:
    # other thing
```
That's a clear sign that it should be two classes.

If there is code to share between the two classes then it can go in a common superclass.

Reply:
Okay, I'll split the code into two different classes.

Reply:
If I split the classes and name them `betainc` and `regularized_betainc`, the problem I face with the latter is that I can't use the `mpmath` function which is named `betainc`. Any way to resolve this? @oscarbenjamin

Reply:
I don't know the implementation of these things so well. Is it not possible to just check the flag before calling into mpmath?

Reply:
The `_eval_evalf` of `Function` tries to match the sympy function with mpmath functions and if there exists a function with the same name, it calls it. The code that does this is here - 
https://github.com/sympy/sympy/blob/1e76752f545a2ed44a8af008c76c9ca7dd77e41d/sympy/core/function.py#L536-L561
An easy way to do this would be to define `_imp_` for fallback evaluation, but that would mean that we would have to import mpmath in the function. I can't think of anything else here

Reply:
Can we not just add this to `MPMATH_TRANSLATIONS`?

Reply:
I think this will do the job, thanks!